pa11ycrawler/reporter.py

```diff
@@ -0,0 +1,260 @@
+                os.makedirs(self.source_dir)
+
+            filepath = os.path.join(
+                self.source_dir, item['filename'] + '.html')
+
+            with open(filepath, 'w') as html_file:
+                html_file.write(item['source'])
+
+    def _update_results_map(self, item):
+        """
+        Adds these results to `self.results_map`.
+        See documentation of `A11yItem` for details about the filename.
+        """
+        self.results_map[item['url']] = item['filename']
+
+    def add_results(self, item):
+        """
+        Saves file to `self.results_dir`.
+        """
+        if not os.path.exists(self.results_dir):
+            os.makedirs(self.results_dir)
+
+        filename = '{}.{}'.format(
+            item['filename'],
+            self.settings.get('PA11Y_REPORTER').replace('-', '.')
+        )
+
+        filepath = os.path.join(self.results_dir, filename)
+        with open(filepath, 'w') as report:
+            report.write(item['results'])
+
+        self._update_results_map(item)
+        self._save_html(item)
+
+    def pre_process_reports(self):
+        """
+        Sets up directories for reports.
+        """
+        if not os.path.exists(self.reports_dir):
+            os.makedirs(self.reports_dir)
+        elif not self.settings.get('PA11YCRAWLER_KEEP_EXISTING_REPORTS'):
+            shutil.rmtree(self.reports_dir)
+
+    def post_process_reports(self):
+        """
+        Saves the `results_map`, which maps a url to the basename of the file
+        that contains its results.
+        """
+        results_map_file = os.path.join(
+            self.reports_dir, 'results', 'pa11y-summary.json')
+
+        with open(results_map_file, 'w') as results_map:
+            json.dump(self.results_map, results_map)
+
+
+class HtmlReporter(ReporterBaseClass):
+    """
+    Generates html reports for the results from Pa11yJsonReporter.
+    """
+    def __init__(self, settings, *args, **kwargs):
+        super(HtmlReporter, self).__init__(settings, *args, **kwargs)
+
+        # where templates are
+        par_dir = os.path.dirname(os.path.abspath(__file__))
+        self.templates_dir = os.path.join(par_dir, 'templates')
+        self.template_lookup = TemplateLookup(directories=[self.templates_dir])
+        self.summary = {
+            "pageResults": {},
+            "overallCount": {
+                "total": 0,
+                "error": 0,
+                "warning": 0,
+                "notice": 0,
+            }
+        }
+
+    def make_html(self):
+        """
+        Main method of this class. Generates HTML reports from the
+        JSON files found in `self.results_dir`.
+        """
+        self._setup_dir()
+        self._copy_assets()
+
+        with open(self.results_map_file, 'r') as report:
+            self.results_map = json.load(report)
+
+        for url, filename in self.results_map.iteritems():
+            result_file = os.path.join(self.results_dir, filename + '.1.0.json')
```

Comment:
The suffix/extension used when saving the data is obtained from the pa11y reporter type, which can be one of html, csv, json, or 1.0-json.  The 1.0 is important because this reporter is built to parse the 1.0-json data.